proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,612 @@
+
+::
+
+   inScope :: Rel Module Name Entity
+   refersTo :: Rel Module QualName Entity
+
+``refersTo`` works as follows:
+
+- If the qualified name has no
+  qualifiers, then it refers to whatever that simple name maps to in the
+  scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifying name refers to. If it refers to a module, then we look at
+  the bindings exported by the module to find one matching the name.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Exporting module names
+^^^^^^^^^^^^^^^^^^^^^^
+
+Module names *cannot* be exported, since there are no export specifiers that can
+refer to them. [#]_ The existing ``module M`` export specifier
+continues to work as it is defined in the Haskell report.
+
+.. [#] See `Module export specifiers`_ for the changes to allow this.
+
+Imports
+^^^^^^^
+
+Imports at the top level of a file module are treated specially. Their
+target is interpreted as a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules. [#]_
+
+.. [#] For imports in other settings, see `Local modules`_.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the target module is no
+brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope with that simple name instead.
+
+If multiple modules are imported ``as`` the same name, then we retain the current
+behaviour, which is to “merge” modules. [#]_ The semantics is roughly that:
+
+.. [#] See `What should happen if multiple modules are defined
+       with the same name?`_ for discussion.
+
+::
+
+   import A as M
+   import B as M
+
+means the same as
+
+::
+
+   module M where
+       import A
+       import B
+
+This retains the current behaviour with respect to clashing names, since
+Haskell does not issue ambiguity warnings on import, only on usage. Thus
+the hypothetical definition of ``M`` is error-free even if ``A`` and
+``B`` export ``n``, but an attempt to use ``M.n`` will result in an
+ambiguity error. [#]_
+
+.. [#] This explication of the semantics uses a local module definition,
+   but we could wire the behaviour in as a special case if we decide
+   not to support local modules.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous section allows us to refer to modules by name, but not to
+export those names.
+
+We propose adding a language extension ``ExportModuleNames`` which
+*changes* the behaviour of the ``module M`` export specifier. It now
```

Comment:
Under this extension, how does one get the old behavior? Would it be `module M(..)`?

Reply:
I haven't included something that provides the old behaviour, but I think `module M(..)` would be a sensible syntax for that.

Reply:
I've added that provisionally along with a note that it needs more thought since I haven't thought about it enough.